src/Common/src/CoreLib/System/IO/PathInternal.cs

```diff
@@ -117,24 +117,54 @@ internal static bool AreRootsEqual(string first, string second, StringComparison
         internal static string RemoveRelativeSegments(string path, int rootLength)
         {
             Debug.Assert(rootLength > 0);
-            bool flippedSeparator = false;
+            
+            Span<char> initialBuffer = stackalloc char[260 /* PathInternal.MaxShortPath */];
+            ValueStringBuilder sb = new ValueStringBuilder(initialBuffer);
+
+            bool flippedSeparator = RemoveRelativeSegmentsHelper(path.AsSpan(), rootLength, ref sb);
+
+            // If we haven't changed the source path, return the original
+            if (!flippedSeparator && sb.Length == path.Length)
+            {
+                sb.Dispose();
+                return path;
+            }
+
+            return sb.Length < rootLength ? path.Substring(0, rootLength) : sb.ToString();
```

Comment:
If sb.Length < rootLength, you're not disposing of the ValueStringBuilder.  Is disposal not needed in that case for some reason?  If so, it'd be worth a comment.  If not, please dispose it.